aten/src/ATen/new_gen.py

```diff
@@ -0,0 +1,871 @@
+        m = re.match(r'^__([^_]+)__$', op)
+        if m is not None:
+            dunder_method = True
+            base = m.group(1)
+            if any(base == f'i{n}' for n in AUGMENTED_ASSIGNMENT_NAMES):
+                inplace = True
+                base = base[1:]
+            else:
+                inplace = False
+                # temporary, this is not intrinsically true but
+                # has been historically true for dunder methods
+                # we support  (but, if we ever got, say, __int__, this would
+                # be wrong!)
+                assert base[0] != 'i'
+        else:
+            dunder_method = False
+            base = op
+            if base[-1] == '_':
+                inplace = True
+                base = base[:-1]
+            else:
+                inplace = False
+        r = BaseOperatorName(base=base, inplace=inplace, dunder_method=dunder_method)
+        assert str(r) == op, f'{str(r)} != {op}'
+        return r
+
+    def __str__(self) -> str:
+        if self.dunder_method:
+            i = 'i' if self.inplace else ''
+            return f'__{i}{self.base}__'
+        else:
+            i = '_' if self.inplace else ''
+            return f'{self.base}{i}'
+
+# Operator name is the base operator name along with the (typically not
+# user visible) overload string.
+@dataclass(frozen=True)
+class OperatorName:
+    name: BaseOperatorName
+    overload_name: str
+
+    @staticmethod
+    def parse(op_name: str) -> 'OperatorName':
+        if '.' in op_name:
+            name, overload_name = op_name.split('.', 1)
+        else:
+            name = op_name
+            overload_name = ''
+        r = OperatorName(
+            name=BaseOperatorName.parse(name),
+            overload_name=overload_name
+        )
+        assert str(r) == op_name, f'{str(r)} != {op_name}'
+        return r
+
+    def __str__(self) -> str:
+        if self.overload_name:
+            return f"{self.name}.{self.overload_name}"
+        else:
+            return f"{self.name}"
+
+# Helper functions for parsing argument lists (both inputs and returns)
+
+def parse_return_arguments(return_decl: str) -> Sequence[Argument]:
+    """
+    Input: '()'
+    Output: []
+    """
+    if return_decl == '()':
+        return []
+    if return_decl[0] == '(' and return_decl[-1] == ')':
+        return_decl = return_decl[1:-1]
+    returns = []
+    for arg in return_decl.split(', '):
+        returns.append(Argument.parse(arg))
+    return returns
+
+def parse_arguments(args: str) -> Tuple[Sequence[Argument], Sequence[Argument], Sequence[Argument]]:
+    """
+    Input: 'int x, int y, int z'
+    Output: positional args, kwarg only args
+    """
+    arguments: List[Argument] = []
+    kwarg_only_arguments: List[Argument] = []
+    out_arguments: List[Argument] = []
+    arguments_acc = arguments
+
+    # TODO: Use a real parser here; this will get bamboozled
```

Comment:
I think a lot of the code in here could benefit from a real parser. There seem to be a few libraries for python parser combinators, they would allow a concise syntax and might be useful.